### ISSUE TITLE
compilation with -Werror=conversion (-DDEV_MODE_STRICT=ON)

### ISSUE THREAD

**Issue by bremond** [siconos/siconos#191]
This is not possible now with a recent gcc. Many, many errors cf:
http://cdash-bipop.inrialpes.fr/viewBuildError.php?buildid=28802

**Comment by radarsat1** [comment]
Two things:

- Is CDash configured to build my unmerged system-csparse branch?  How?  Was it automatic or something someone changed?  Is there control for this somewhere?  I don't actually see the branch name anywhere but that clearly has the CXSparse sources so it must be my upgrade branch which hasn't even been pushed to the main repository yet.  Not complaining, good to see CDash results in advance, but just to be clear that this is not yet in master.

- I really doubt it would be a good use of time to fix conversion warnings that come from sources external to Siconos.  Modifying dependencies just makes it hard to update them in the future.  I propose disabling this for `externals`.  (Warnings are fine.)

**Comment by radarsat1** [comment]
I was just looking a little closer and it seems there have been problems in master since at least September.  I will merge system-csparse so that we can work on top of that, as it is sufficiently mixed up with NumericsMatrix etc.

**Comment by radarsat1** [comment]
Some problem classes.

## External code:
~~~
/home/sinclairs/projects/siconos/externals/sort/sort_common.h:77:22: error: conversion to ‘int’ from ‘uint64_t {aka const long unsigned int}’ may alter its value [-Werror=conversion]                                                                                   
   const int minrun = size >> shift;                                                     
                      ^~~~                                                               
In file included from /home/sinclairs/projects/siconos/numerics/src/tools/NumericsSparseMatrix.c:52:0: 
~~~

which means that there are errors in this mode that come from headers in externals.  Meaning that even if we omit `-Werror-conversion` for Externals, we will get these errors in Numerics.
 
So the question is, do we want to start fixing these kinds of errors in external code, or leave them?  It's an unfortunate choice to have to make.

## SWIG-generated code:
~~~
/home/sinclairs/projects/siconos/bldpy2/wrap/siconos/numericsPYTHON_wrap.c: In function ‘SWIG_AsVal_double’:
/home/sinclairs/projects/siconos/bldpy2/wrap/siconos/numericsPYTHON_wrap.c:3418:21: error: conversion to ‘double’ from ‘long int’ may alter its value [-Werror=conversion]
     if (val) *val = PyInt_AsLong(obj);
                     ^~~~~~~~~~~~
~~~

This function `SWIG_AsVal_double`, is a hard-coded part of SWIG.  We may have to exclude swig-generated files from `-Werror=conversion`.

## Index representation, dependencies on BLAS etc

There are also some situations where I tried changing the use of `int` to `size_t` in order to start adapting code to a uniform index type, and came across situations where it was impossible to avoid down-converting, e.g., in `NM_row_prod_no_diag3` calling `blas_ddot` takes `int` even though we have `size_t` as the function arguments.

The opposite solution sort of seems to work, but we can no longer use `size_t`.  In that case, possibly there is no point supporting SICONOS_INT64.

Although it's good to catch these conversion errors because it means that even if we use `size_t` as an index, we effectively don't support 64-bit indexes since they are down-converted to 32-bit in some places.  It would be better to uniformly convert to 64-bit integers (`size_t`) but as long as we use functions in BLAS etc that don't support it, we cannot.

In other words, treating these as errors is absolutely right.  Not that I expect to be working with such large arrays, but if we pretend to support 64-bit indexing then it should work across the board.

**Comment by xhub** [comment]
- External : disable the error and even warning. I remember fixing some of those issues in our customized copy of csparse, or making the cast explicit. It's not our codebase.
- For swig, we could try to exclude the SWIG part, and enforce those flags just for the code we write. Or we could exclude those flags. It would be nice to do a PR for swig.
- for blas and other linear algebra library, an ever increasing number of then support 32 bits and 64 bits. OpenBLAS can be build as a 64 bits BLAS library. size_t may not be the right type, since it is unsigned, and I believe the linear algebra library only support signed integer (to be checked). I would say that the implicit conversion from int64_t to int32_t is bad, but I would not worry too much. We could try to enforce a check during the creation of the matrix ( NM_create and NM_create_from_data) that the size does not exceed the supported length (or at least an assert). I see this as the user not providing the right kind of library to tackle his large scale problem. The main issue is when there are array of integers: in that case, the result is wrong for almost all problem sizes, and we have to take care of that aspect and ensure correctness.

Note that with my GCC (7.2.0), I have the following flags: -Wconversion, -Wsign-conversion and -Wfloat-conversion
I believe we already disable -Wsign-conversion whenever it is a supported flag. Maybe it is easier to enforce Werror=float-conversion in some places?

**Comment by radarsat1** [comment]
Agreed on all points. But more immediately, what do we do about CDash?  As far as I can tell there builds with 0 errors and builds with 50+ errors.  It's very confusing to read, and makes most of the test environments useless as they merely repeat the same type conversion errors.  Depending on the importance of this issue, I would suggest leaving DEV_MODE_STRICT=OFF for most builds, and only enabling it for one or two environments with "everything" turned on.

**Comment by xhub** [comment]
I'll try to come up with a fix soon. My guess is that the newer GCC needs to not have certain error flags.

Also most of those issue come from cxsparse? We should by default install it systemwide and have one or two config without.
 
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by bremond** [comment]
normally commit on branches should result in a built (this is the config in
Jenkins) and normally the name of the branch should be in the name of
the build (it works for the last release 4.1.x). Well, this is what I
would like to have.

But something is not working...
Maybe these builds come from travis and the branch name fails with
travis.


Note that you can check the sha1 of the sources in the Notes, for
example:
http://cdash-bipop.inrialpes.fr/viewNotes.php?buildid=28801

which is reachable from
http://cdash-bipop.inrialpes.fr/buildSummary.php?buildid=28801

I did not check, I hope this sha1 is the good one.

>I really doubt it would be a good use of time to fix conversion
>warnings that come from sources external to Siconos.
I totally agree. I think we should not have -Werror=conversion on
externals and that we should have only one config (for the moment) on
cdash that show -Werror=conversion on siconos internals.

**Comment by bremond** [comment]
in c3ea4b10d, I removed the DEV_MODE_STRICT from default and created a
with_dev_mode_strict config for CI. I suggest that we keep this config, even if
Werror=conversion are all fixed in internal, in order to have a place to
put other strict checks that are not immediate to fix.

**Comment by radarsat1** [comment]
>  Also most of those issue come from cxsparse?

A bunch of them are also in NumericsMatrix.c and a few other files.  Actually I really think these look like they would help fixing potential issues with indexing of large matrices, it's just that they shouldn't block the purpose of the other builds.  Especially since as it appears (and I've experienced in the past) that new versions of compilers introduce new warnings, and so if they are treated as errors, suddenly your code doesn't compile...  I think @bremond's change that he describes is welcome for that purpose.
